File: 05-Algoritmos/SistemaHUB.js
class Dispositivo {
    constructor(nome) {
        this.nome = nome;
    }

    atualizar(dados) {
        console.log(`Dispositivo ${this.nome} recebeu dados: ${dados}`);
    }
}

class Hub {
    constructor(nome, maxPortas) {
        this.nome = nome;
        this.maxPortas = maxPortas;
        this.dispositivos = [];
    }

    conectarDispositivo(dispositivo) {
        if (this.dispositivos.length < this.maxPortas) {
            this.dispositivos.push(dispositivo);
            console.log(`${dispositivo.nome} conectado ao ${this.nome}`);
        } else {
            console.log("Limite de portas atingido!");
        }
    }

    desconectarDispositivo(dispositivo) {
        this.dispositivos = this.dispositivos.filter(d => d !== dispositivo);
        console.log(`${dispositivo.nome} desconectado do ${this.nome}`);
    }

    transmitirDados(dados, dispositivoDestino, dispositivoOrigem) {
        console.log(`Transmitindo dados de ${dispositivoOrigem.nome} para ${dispositivoDestino.nome}`);
        this.dispositivos.forEach(dispositivo => {
            if (dispositivo !== dispositivoOrigem) {
                dispositivo.atualizar(dados);
            }
        });
    }
}

// Exemplos de uso da classe Hub
let hub = new Hub("Hub Principal", 4);
let dispositivo1 = new Dispositivo("Dispositivo A");
let dispositivo2 = new Dispositivo("Dispositivo B");
hub.conectarDispositivo(dispositivo1);
hub.conectarDispositivo(dispositivo2);
hub.transmitirDados("Mensagem de teste", dispositivo2, dispositivo1);

File: 05-Algoritmos/sistema_hub.py
# Simulação de um Hub em Python

class Hub:
    def __init__(self, nome, max_portas):
        self.nome = nome
        self.max_portas = max_portas
        self.dispositivos = []

    def conectar_dispositivo(self, dispositivo):
        if len(self.dispositivos) < self.max_portas:
            self.dispositivos.append(dispositivo)
            print(f"{dispositivo} conectado ao {self.nome}")
        else:
            print("Limite de portas atingido!")

    def desconectar_dispositivo(self, dispositivo):
        if dispositivo in self.dispositivos:
            self.dispositivos.remove(dispositivo)
            print(f"{dispositivo} desconectado do {self.nome}")

    def transmitir_dados(self, dados, dispositivo_destino, dispositivo_origem):
        print(f"Transmitindo dados de {dispositivo_origem} para todos os dispositivos conectados (exceto origem)")
        for dispositivo in self.dispositivos:
            if dispositivo != dispositivo_origem:
                if dispositivo == dispositivo_destino:
                    print(f"Dados recebidos na porta de {dispositivo}: {dados}")
                else:
                    print(f"Replicando dados para {dispositivo}")

    def dispositivos_conectados(self):
        return self.dispositivos

# Exemplos de uso
hub = Hub("Hub Principal", 4)
dispositivo1 = "Dispositivo A"
dispositivo2 = "Dispositivo B"
hub.conectar_dispositivo(dispositivo1)
hub.conectar_dispositivo(dispositivo2)
hub.transmitir_dados("Mensagem de teste", dispositivo2, dispositivo1)